UnitTests/BraintreeVisaCheckoutTests/BTVisaCheckoutClient_Tests.swift

```diff
@@ -1,47 +1,470 @@
+            guard let error = error as NSError? else {
+                XCTFail("Expected an error to be returned")
+                return
+            }
+            XCTAssertEqual(error.code, BTVisaCheckoutError.checkoutUnsuccessful.rawValue)
+            expectation.fulfill()
+        }
+
+        self.waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenStatusCodeIndicatesCancellation_callsAnalyticsWithCancelled() {
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expectation = self.expectation(description: "Analytic sent")
+        
+        client.tokenize(statusCode: .statusUserCancelled, callID: callID, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { _, _ in
+            XCTAssertEqual(self.mockAPIClient.postedAnalyticsEvents.last, BTVisaCheckoutAnalytics.tokenizeFailed.description)
+            expectation.fulfill()
+        }
+
+        self.waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenStatusCodeIndicatesError_callsAnalyticsTokenizeFailed() {
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expectation = self.expectation(description: "Analytic sent")
+
+        client.tokenize(statusCode: .statusInternalError, callID: callID, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { _, _ in
+            XCTAssertEqual(self.mockAPIClient.postedAnalyticsEvents.last, BTVisaCheckoutAnalytics.tokenizeFailed)
+            expectation.fulfill()
+        }
+        self.waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenTokenizationErrorOccurs_callsCompletionWithError() {
+        mockAPIClient.cannedConfigurationResponseBody = BTJSON(value: [
+            "environment": "sandbox",
+            "visaCheckout": [
+                "apikey": "API Key",
+                "externalClientId": "clientExternalId",
+                "supportedCardTypes": [
+                    "Visa",
+                    "MasterCard",
+                    "American Express",
+                    "Discover"
+                ]
+            ]
+        ])
+
+        mockAPIClient.cannedHTTPURLResponse = HTTPURLResponse(url: URL(string: "any")!, statusCode: 503, httpVersion: nil, headerFields: nil)
+        mockAPIClient.cannedResponseError = NSError(domain: "https://braintree.com", code: 123, userInfo: nil)
+
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expecation = expectation(description: "tokenization error")
+        
+        client.tokenize(statusCode: .statusSuccess, callID: callID, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { nonce, error in
+            if nonce != nil {
+                XCTFail("Expected nonce to be nil")
+                return
+            }
```

Comment:
Can we instead assert that the nonce is nil?

Reply:
`XCTAssertNil` is specifically designed for validating the absence of a value in an optional type, which is the expected behavior of our tests. `XCTFail`  is used to unconditionally fail a test at a specific point in the code. It does not evaluate any expression; it simply signals a test failure. In these cases where we check for `!= nil` we should actually be testing that the optional value we expect to be nil is nil instead, so `XCTAssertNil` is more appropriate.